src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,184 @@
+    end
+
+let (<++>) = trace_concat
+let (<-->) = trace_subtract
+
+/// Traces form a monoid with trace concatenation as operation, and Nil as unit.
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (
+      tr <++> empty_trace == tr /\
+      empty_trace <++> tr == tr
+    )
+    [SMTPatOr [[SMTPat (tr <++> empty_trace)]; [SMTPat (empty_trace <++> tr)]]]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (append_entry tr e == (tr <++> (Snoc Nil e)))
+    [SMTPat (tr <++> (Snoc Nil e))]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t)
+
+let rec trace_concat_grows (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (tr1 <$ (tr1 <++> tr2))
+    [SMTPat (tr1 <$ (tr1 <++> tr2))]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr2 with
+    | Nil -> ()
+    | Snoc hd e -> trace_concat_grows tr1 hd
+
+/// Properties of trace subtraction
+
+let trace_subtract_append_entry (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((append_entry tr2 e) <--> tr1) == append_entry (tr2 <--> tr1) e)
+    [SMTPat ((append_entry tr2 e) <--> tr1)]
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t)
+
+let rec trace_subtract_trace_length (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures trace_length (tr2 <--> tr1) == trace_length tr2 - trace_length tr1)
+    [SMTPat (trace_length (tr2 <--> tr1))]
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t);
+    if trace_length tr2 = trace_length tr1 then ()
+    else begin
+      grows_cases tr1 tr2;
+      trace_subtract_trace_length tr1 (init tr2)
+    end
+
+let rec trace_subtract_get_entry (#label_t:Type) (tr1 tr2:trace_ label_t) (i:timestamp{i `on_trace` tr2})
+  : Lemma
+    (requires tr1 <$ tr2 /\ i >= trace_length tr1)
+    (ensures get_entry_at tr2 i == get_entry_at (tr2 <--> tr1) (i - trace_length tr1))
+  = norm_spec [zeta; delta_only [`%trace_subtract]] (trace_subtract #label_t);
+    if i = last_ts tr2 then ()
+    else begin
+      grows_cases tr1 tr2;
+      trace_subtract_get_entry tr1 (init tr2) i
+    end
+
+let rec trace_subtract_fmap_trace (#a #b:Type) (tr1 tr2:trace_ a) (f:a -> b)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures (fmap_trace f (tr2 <--> tr1) == ((fmap_trace f tr2) <--> (fmap_trace f tr1))))
```

Comment:
There is no such lemma for concatenation?

Reply:
Indeed there could be, though one wasn't yet needed. I'll add corresponding concat versions of the subtraction lemmas.